src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
```

Comment:
maybe some of the cost that was measured/perceived in Roslyn, where a similar technique is used with recursion depth tracking, comes from the fact that `EnsureSufficientExecutionStack` is accessed through a portable shim, thus involving delegate invocations:
```csharp
        internal static class RuntimeHelpers
        {
            internal const string TypeName = "System.Runtime.CompilerServices.RuntimeHelpers";

            internal static readonly Type Type = ReflectionUtilities.GetTypeFromEither(
                contractName: $"{TypeName}, {CoreNames.System_Runtime}",
                desktopName: TypeName);

            internal static readonly Action EnsureSufficientExecutionStack = Type
                .GetTypeInfo()
                .GetDeclaredMethod(nameof(EnsureSufficientExecutionStack), paramTypes: new Type[] { })
                .CreateDelegate<Action>();
        }
```
When updating my micro-benchmark to incorporate this option as well, I can see a very small additional cost due to the delegate indirection, and omitting the threshold management and check still yields marginally better numbers:
```
Direct / no threshold:   00:00:13.3524722
Direct / threshold:      00:00:14.0316492
Delegate / no threshold: 00:00:13.6635413
Delegate / threshold:    00:00:14.1541278
```
This is with /o+ on x64 on Windows 10 on Release builds. My vote is to take out the threshold check and recursion depth tracking logic.